### ISSUE TITLE
Authorization:[Basic xxxxx] Header is not passed  to upstream 

### ISSUE THREAD

**Issue by morarucostel** [oauth2-proxy/oauth2-proxy#394]
In our scenario, we are using the basic-auth of `oauth2_proxy` to authenticate users against the htpasswd file. 
It is deployed as an Docker image in a kubernetes cluster and the secured application is accessed through ingress and the controller is done through NGINX.

Once the authentication is done successfully and the flow reaches `addHeadersForProxying`, the oauth-proxy is setting-up correctly the Authorization (to Basic) and X-Forwarded-User headers. However the header doesn't reach the upstream applications even though in the NGINX snippet we have 
```
      auth_request_set $token $upstream_http_authorization;
      proxy_set_header Authorization $token;
```

## Expected Behavior

The upstream applications should receive the Authorization: Basic header.

## Steps to Reproduce (for bugs)

The args to start the oauth-proxy are:
```
  - --email-domain=SPECIFIC_VALUE
  - --cookie-secure=true
  - --cookie-domain=SPECIFIC_VALUE
  - --cookie-path=/SPECIFIC_VALUE
  - --upstream=file:///dev/null
  - --http-address=0.0.0.0:4180
  - --cookie-secret=SPECIFIC_VALUE
  - --scope=openid
  - --skip-provider-button=false
  - --pass-access-token=true
  - --htpasswd-file=/opt/config/htpasswd
  - --cookie-refresh=0
  - --pass-basic-auth=true
```

The ingress definition with the NGINX snippet is:
```
    ingress.kubernetes.io/configuration-snippet: |
      proxy_pass_request_headers on;
      auth_request_set $token $upstream_http_authorization;
      proxy_set_header Authorization $token;
```

After the successful authentication, even thought the Authorization header is set in the code, it doesn't get propagated to the upstream service.

## Your Environment

- Version used: 4.1.0

**Comment by JoelSpeed** [comment]
Hey @morarucostel could you please confirm which headers it is that you are expecting your upstream application to receive?

**Comment by tlawrie** [comment]
Hey @JoelSpeed it is the Authorization header with the "Basic <username:password>" that we are looking for. But it doesn't seem to make it to the backend systems.

**Comment by JoelSpeed** [comment]
Have you tried using the `nginx.ingress.kubernetes.io/auth-response-headers` annotation that nginx-ingress provides?

**Comment by morarucostel** [comment]
Hey @JoelSpeed nope, not even with the `nginx.ingress.kubernetes.io/auth-response-headers` annotation.

**Comment by ploxiln** [comment]
You're trying to get an `Authorization` header from the auth-request response, but it is not a response header, it is a request header for upstream requests in proxy mode. If you enable `--set-xauthrequest` then you will get the `X-Auth-Request-User` response header which you can access as `$upstream_http_x_auth_request_user`.

... but do you actually want the basic auth that was passed to oauth2_proxy in the original request, to also be passed to the upstream? In that case I think you can just not try to get it from the oauth2_proxy response and not replace the Authorization header in the request sent to the upstream app.

**Comment by tlawrie** [comment]
hey @ploxiln it worked to get the user using that method but we are wanting the whole Authorization header.

We can see the auth proxy is setting it (we added extra logging to see all the headers) however using the same sort of logic for the Authorization header
`proxy_set_header Authorization $http_authorization;`

We also used the annotation mentioned by @JoelSpeed and documented on nginx ingress controller. It automatically added
`auth_request_set $authHeader0 $upstream_http_authorization;
proxy_set_header 'Authorization' $authHeader0;`

But that doesn't come through to our backend service either... any further thoughts on what might be interrupting this?

**Comment by ploxiln** [comment]
It looks like there is one place where Authorization is set as a *response* header for the auth request if you enable `--set-authorization-header`, but it only works for oauth tokens, not for basic auth:

```go
	if p.SetAuthorization {
		if session.IDToken != "" {
			rw.Header().Set("Authorization", fmt.Sprintf("Bearer %s", session.IDToken))
```

Contrast it to where the basic auth is set on the proxied request (which is not used in auth-response mode) (notice `req` vs `rw`)

```go
	if p.PassBasicAuth {
		req.SetBasicAuth(session.User, p.BasicAuthPassword)
		req.Header["X-Forwarded-User"] = []string{session.User}
```

**Comment by morarucostel** [comment]
@ploxiln Yes, that is the problem. 
There is now way in setting the Basic Authorization header to the _response_ headers.

Keeping consistent with `set` vs `pass` shouldn't we have also a `-set-basic-auth` option that would set the Basic Authorization header on the response?

Something like 
```
	if p.SetBasicAuth {
		if session.User != "" {
			rw.Header().Set("Authorization", fmt.Sprintf("Basic %s", base64.StdEncoding.EncodeToString([]byte(session.User+":"+p.BasicAuthPassword))))
		} else {
			rw.Header().Del("Authorization")
		}
	}
```
 
that would be here.
https://github.com/pusher/oauth2_proxy/blob/bd79b976daddb753c18f86e6bf6764b60ecc80f2/oauthproxy.go#L932

I can raise a PR if you want me to.

**Comment by morarucostel** [comment]
Issue closed based on the #413 PR.
